core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,324 @@
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.internal.common.JacksonUtil;
+
+/**
+ * Generates a JSON Schema from the given service specification.
+ */
+final class JsonSchemaGenerator {
+
+    private static final ObjectMapper mapper = JacksonUtil.newDefaultObjectMapper();
+
+    /**
+     * Generate an array of json schema specifications for each method inside the service.
+     * @param serviceSpecification
+     * @return ArrayNode that contains service specifications
+     */
+    public static ArrayNode generate(ServiceSpecification serviceSpecification) {
+        final JsonSchemaGenerator generator = new JsonSchemaGenerator(serviceSpecification);
+        return generator.generate();
+    }
+
+    private final Set<EnumInfo> enumInfos;
+    private final Set<StructInfo> structInfos;
+    private final Set<ServiceInfo> serviceInfos;
+    private final Map<String, StructInfo> typeNameToStructMapping;
+    private final Map<String, EnumInfo> typeNameToEnumMapping;
+
+    private JsonSchemaGenerator(ServiceSpecification serviceSpecification) {
+        this.enumInfos = serviceSpecification.enums();
+        this.structInfos = serviceSpecification.structs();
+        this.serviceInfos = serviceSpecification.services();
+        this.typeNameToStructMapping = structInfos.stream().collect(
+                Collectors.toMap(StructInfo::name, Function.identity()));
+        this.typeNameToEnumMapping = enumInfos.stream().collect(
+                Collectors.toMap(EnumInfo::name, Function.identity()));
+    }
+
+    private ArrayNode generate() {
+        final ArrayNode definitions = mapper.createArrayNode();
+
+        final Set<ObjectNode> methodDefinitions = serviceInfos.stream().flatMap(
+                serviceInfo -> serviceInfo.methods().stream()
+                                          .map(methodInfo -> generate(serviceInfo, methodInfo))).collect(
+                Collectors.toSet());
+
+        return definitions.addAll(methodDefinitions);
+    }
+
+    /**
+     * Generate the JSON Schema for the given {@link ServiceInfo} and {@link MethodInfo}.
+     * @param serviceInfo service info.
+     * @param methodInfo the method to generate the JSON Schema for.
+     * @return ObjectNode containing the JSON schema for the parameter type.
+     */
+    private ObjectNode generate(ServiceInfo serviceInfo, MethodInfo methodInfo) {
+        final ObjectNode root = mapper.createObjectNode();
+
+        root.put("$id", methodInfo.id())
+            .put("title", methodInfo.name())
+            .put("description", methodInfo.descriptionInfo().docString())
+            // TODO: Assumes every method takes an object, which is only valid for RPC based services
+            //  and most of the REST services.
+            .put("additionalProperties", false).put("type", "object");
```

Comment:
How about defining each struct in `$defs` section and link to it using `$refs` instead of putting the information in place?
We can create a map that maps the `structInfo.name()` with the path that we are going to put them in the `$defs` section and link to it using:
```java
methodInfo.parameters().forEach(parameter -> {
    final TypeSignature typeSignature = parameter.typeSignature();
    switch (typeSignature.type()) {
        case BASE:
            root.set(parameter.name(),
                     mapper.createObjectNode().put("type", getSchemaType(typeSignature)));
            break;
        case STRUCT:
            root.set(parameter.name(),
                     mapper.createObjectNode().put("$ref", reference(typeSignature))); // This will look up the map.
            break;
        case ENUM:
            ...
    }
});
```

Reply:
Defining each struct in `$def` seems like a good idea. We can also break down our schema-related endpoint into two:
- An endpoint that returns schemas for all methods excluding `$def`
- An endpoint that returns `$def` 

And frontend can simply inject that def section to each method when using, avoiding duplication. This makes sense from DocService's perspective, for other users perspective having redundant fields in $def might be annoying so maybe we can use relative URI s in $ref which I think is a little more error-prone.

Reply:
When you maintain those schemas by hand `$def` seems nice but when auto-generated I feel like seeing everything on a single place is better 🙂 The reason is that we can just use an IDE to collapse fields while reviewing which would make the layout look a lot like `$defs` structure without users needing constantly jumping around references just by looking. WDY Think?

Reply:
> when auto-generated I feel like seeing everything on a single place is better 🙂 The reason is that we can just use an IDE to collapse fields while reviewing which would make the layout look a lot like $defs structure without users needing constantly jumping around references just by looking.

That also makes sense. I was just worried about the case where a big struct is used in so many places which causes extra bytes for transmission.

Let's do what you suggested which put the structs in place and change later if there's a demand for that. 😄

Reply:
So we can probably use the same strategy that pre instantiates the struct object and store them in a map and use them in place.

Reply:
Oops, I realized that pre instantiating won't work  due to the nested structs. We construct the struct in place using the type